File: Java/LeetCode/Array/DuplicateZeros.py
class Solution:
    def duplicateZeros(self, arr: list[int]) -> None:
        """
        Do not return anything, modify arr in-place instead.
        """
        i = 0;
        while i < len(arr):
            if (arr[i] == 0):
                for j in range(len(arr) - 1, i, -1):
                    arr[j] = arr[j - 1]
                i += 1 # skip the shifted 0
            i += 1
        print(arr)


def main():
    sol = Solution()
    arr = [1,0,2,3,0,4,5,0]
    sol.duplicateZeros(arr)
    arr = [1,2,3]
    sol.duplicateZeros(arr)

        
if __name__ == '__main__':
    main()

File: Java/LeetCode/Array/DuplicateZeros.java
// LC 1089 -- Array
public class DuplicateZeros {
    public void duplicateZeros(int[] arr) {
        int numOfZero = 0;
        for (int i : arr) {
            if (i == 0) {
                numOfZero++;
            }
        }
        int n = arr.length;
        for (int i = n - 1; i >= 0; i--) {
            if (i + numOfZero < n) { 
                arr[i + numOfZero] = arr[i];  
            }

            if (arr[i] == 0) {
                numOfZero--;
                if (i + numOfZero < n) {
                    arr[i + numOfZero] = 0;
                }
            } 
        }
    }

}
